### PR TITLE
Automatic name resolution in declareProperty

### LINKED ISSUE #15956: Automatic name resolution for declareProperty
## Reason
getMaterialProperty, and similar syntax, allow for automatic name recognition from the input parameters, such that a name supplied in the InputParams,

`  params.addRequiredParam<MaterialPropertyName>("property_name",
                                                "Name of porosity material property");`

is automatically picked up in the constructor,

```
    _property_old(getMaterialPropertyOld<Real>("property_name"))
```

Declare properties do not have the same name resolution, requiring, 
```
_property(declareProperty<Real>(getParam<MaterialPropertyName>("property_name"))),
```

## Design
Allow for automatic name resolution with `declareProperty`, `declareADProperty`, and` declareGenericProperty`.

## Impact
While this is somewhat minor, it has led to problems when users (like myself) have assumed that declare properties has automatic name resolution.

### PR DESCRIPTION
**Pull request by tophmatthews**
These tests are provided for test driven design for better name handling with `declareProperty`, `declareADProperty`, and declare `declareGenericProperty`. ~~They do not include a fix at this time~~.

Closes #15956

### CODE REVIEW

**Comment by tophmatthews** [comment]
This happens for both `GenericStateful` and `ADGenericStateful`.

**Comment by tophmatthews** [comment]
What am I screwing up here in the AD/non-AD stateful material here?

**Comment by tophmatthews** [comment]
<details>
  <summary>Full assert </summary>
  ```
Time Step 0, time = 0

Time Step 1, time = 1, dt = 1


Assertion `i < _size' failed
Access out of bounds in MooseArray (i: 4 size: 4)
at /Users/cmatthews/projects/moose/framework/build/header_symlinks/MooseArray.h, line 279
Stack frames: 28
0: 0   libmesh_dbg.0.dylib                 0x0000000110182c2c libMesh::print_trace(std::__1::basic_ostream<char, std::__1::char_traits<char> >&) + 460
1: 1   libmoose_test-dbg.0.dylib           0x00000001086838cd MooseArray<double>::operator[](unsigned int) + 941
2: 2   libmoose_test-dbg.0.dylib           0x00000001086f5b84 MaterialPropertyBase<double, false>::operator[](unsigned int) + 36
3: 3   libmoose_test-dbg.0.dylib           0x000000010884a935 GenericStatefulTempl<false>::computeProperties() + 37
4: 4   libmoose-dbg.0.dylib                0x000000010a0e022c MaterialData::reinit(std::__1::vector<std::__1::shared_ptr<MaterialBase>, std::__1::allocator<std::__1::shared_ptr<MaterialBase> > > const&, bool) + 172
5: 5   libmoose-dbg.0.dylib                0x000000010a59ebe7 FEProblemBase::reinitMaterials(unsigned short, unsigned int, bool) + 503
6: 6   libmoose-dbg.0.dylib                0x000000010ac1a1d4 ComputeResidualThread::onElement(libMesh::Elem const*) + 276
7: 7   libmoose-dbg.0.dylib                0x000000010a45b768 ThreadedElementLoopBase<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> >::operator()(libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> const&, bool) + 552
8: 8   libmoose-dbg.0.dylib                0x000000010acdbbbc void libMesh::Threads::parallel_reduce<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*>, ComputeResidualThread>(libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> const&, ComputeResidualThread&) + 156
9: 9   libmoose-dbg.0.dylib                0x000000010acd471f NonlinearSystemBase::computeResidualInternal(std::__1::set<unsigned int, std::__1::less<unsigned int>, std::__1::allocator<unsigned int> > const&) + 863
10: 10  libmoose-dbg.0.dylib                0x000000010acd3d79 NonlinearSystemBase::computeResidualTags(std::__1::set<unsigned int, std::__1::less<unsigned int>, std::__1::allocator<unsigned int> > const&) + 617
11: 11  libmoose-dbg.0.dylib                0x000000010a5b8b5a FEProblemBase::computeResidualTags(std::__1::set<unsigned int, std::__1::less<unsigned int>, std::__1::allocator<unsigned int> > const&) + 2346
12: 12  libmoose-dbg.0.dylib                0x000000010a5b7ca6 FEProblemBase::computeResidualInternal(libMesh::NumericVector<double> const&, libMesh::NumericVector<double>&, std::__1::set<unsigned int, std::__1::less<unsigned int>, std::__1::allocator<unsigned int> > const&) + 230
13: 13  libmoose-dbg.0.dylib                0x000000010a5b79dc FEProblemBase::computeResidual(libMesh::NumericVector<double> const&, libMesh::NumericVector<double>&) + 252
14: 14  libmoose-dbg.0.dylib                0x000000010a5b77e6 FEProblemBase::computeResidualSys(libMesh::NonlinearImplicitSystem&, libMesh::NumericVector<double> const&, libMesh::NumericVector<double>&) + 102
15: 15  libmoose-dbg.0.dylib                0x000000010acc9b86 NonlinearSystem::solve() + 390
16: 16  libmoose-dbg.0.dylib                0x000000010a5b53e6 FEProblemBase::solve() + 374
17: 17  libmoose-dbg.0.dylib                0x000000010a082694 FEProblemSolve::solve() + 36
18: 18  libmoose-dbg.0.dylib                0x000000010a088513 PicardSolve::solveStep(double, double&, double, double&, bool, std::__1::set<unsigned int, std::__1::less<unsigned int>, std::__1::allocator<unsigned int> > const&) + 1619
19: 19  libmoose-dbg.0.dylib                0x000000010a086987 PicardSolve::solve() + 3687
20: 20  libmoose-dbg.0.dylib                0x000000010aadd0bc TimeStepper::step() + 44
21: 21  libmoose-dbg.0.dylib                0x000000010a08e8c0 Transient::takeStep(double) + 288
22: 22  libmoose-dbg.0.dylib                0x000000010a08dfcb Transient::execute() + 171
23: 23  libmoose-dbg.0.dylib                0x000000010b236dc8 MooseApp::executeExecutioner() + 232
24: 24  libmoose-dbg.0.dylib                0x000000010b237d0f MooseApp::run() + 335
25: 25  moose_test-dbg                      0x0000000108659a15 main + 149
26: 26  libdyld.dylib                       0x00007fff6f51dcc9 start + 1
27: 27  ???                                 0x0000000000000003 0x0 + 3
[0] /Users/cmatthews/projects/moose/framework/build/header_symlinks/MooseArray.h, line 279, compiled Oct 15 2020 at 15:54:40
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
```
</details>

**Comment by tophmatthews** [comment]
Alright, I added more more test objects and inputs. I have an AD stateful material, a Regular stateful material, an AD stateful material from the templated combined methods, and a regular stateful material from the template combined methods. All fail in dbg on MacOS. The regular stateful fails intermittently on opt...I must be screwing up something bad somewheres...

File: test/src/materials/Stateful.C [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+void
+Stateful::initQpStatefulProperties()
+{
+  _property[_qp] = 0.0;
+}
+
+void
+Stateful::computeProperties()
```
Comment (tophmatthews) [inline test/src/materials/Stateful.C:39]: `computeQpProperties`...🤮
oh man, it was a long week when I put this in...

File: framework/include/materials/MaterialBase.h [hunk 2]
```diff
@@ -341,7 +346,12 @@ template <typename T>
...
+  // Check if the supplied parameter is a valid imput parameter key
+  std::string mat_name = prop_name;
+  if (_pars.have_parameter<MaterialPropertyName>(prop_name))
+    mat_name = _pars.get<MaterialPropertyName>(prop_name);
+
   _fe_problem.usingADMatProps(true);
-  registerPropName(prop_name, false, MaterialBase::CURRENT);
-  return materialData().declareADProperty<T>(prop_name);
+  registerPropName(mat_name, false, MaterialBase::CURRENT);
+  return materialData().declareADProperty<T>(mat_name);
```
Comment (dschwen) [inline framework/include/materials/MaterialBase.h:356]: prop_name -> name

mat_name -> prop_name

please

**Review by dschwen — CHANGES_REQUESTED** [review]

File: framework/include/materials/MaterialBase.h (hunk 2 above)
Comment (dschwen) [reply framework/include/materials/MaterialBase.h:356]: The input parameter is either a property name of an input parameter name, let's cal it `name`. A material name is not the same as a material property name, `prop_name` is the right thing here.

Comment (tophmatthews) [reply framework/include/materials/MaterialBase.h:356]: so do you want `declareADProperty(const std::string & name)`?

Comment (tophmatthews) [reply framework/include/materials/MaterialBase.h:356]: or just mat_name -> name?

Comment (tophmatthews) [reply framework/include/materials/MaterialBase.h]: I did the latter since it changes less things, let me know if it makes sense to you

**Comment by tophmatthews** [comment]
Well, I cannot figure out where the material dependency comes from. There is something screwy with the damage models @bwspenc or @dschwen  that you may want to really make sure are doing what you think they are doing...

**Comment by tophmatthews** [comment]
The error is 
```
The following material properties are declared on block 0 by multiple materials:
  Material Property             Material Objects
  damage_index_prop             all_strain damage damage_index elasticity strain_xx_all stress stress_xx_all 
```
I think it may be due to something with the old material values, but I can't really tell. Things are not being passed around nicely though.

**Comment by tophmatthews** [comment]
Aie Chihuahua, it's a consequence of this line:
```
std::string name = prop_name;
  if (_pars.have_parameter<MaterialPropertyName>(prop_name))
    name = _pars.get<MaterialPropertyName>(prop_name);
```
In the case of the `ScalarDamageBase` failure, when declaring the damage index property:
```
_damage_index_name(getParam<MaterialPropertyName>("damage_index_name")),
_damage_index(declareProperty<Real>(_base_name + _damage_index_name)),
```
equates to
```
_damage_index(declareProperty<Real>("damage_index")),
```
when there is no base_name, and _damage_index_name defaults to "damage_index" in the input parameters.

Now in the inherited class `ScalarMaterialDamage`, there is a `MaterialPropertyName` input parameter "damage_index". So `ScalarDamage` picks up the name of that material and double creates a material with that name...

File: framework/include/materials/MaterialBase.h [hunk 3]
```diff
@@ -280,6 +284,18 @@ class MaterialBase : public MooseObject,
...
 MaterialBase::declareProperty(const std::string & prop_name)
+{
+  // Check if the supplied parameter is a valid input parameter key
+  std::string name = prop_name;
+  if (_pars.have_parameter<MaterialPropertyName>(prop_name))
+    name = _pars.get<MaterialPropertyName>(prop_name);
+
+  return declarePropertyByName<T>(name);
+}
+
```
Comment (dschwen) [inline framework/include/materials/MaterialBase.h:295]:
```suggestion
MaterialBase::declareProperty(const std::string & name)
{
  // Check if the supplied parameter is a valid input parameter key
  std::string prop_name = name;
  if (_pars.have_parameter<MaterialPropertyName>(name))
    prop_name = _pars.get<MaterialPropertyName>(name);

  return declarePropertyByName<T>(prop_name);
}

```

Comment (dschwen) [reply framework/include/materials/MaterialBase.h:295]: I thought this is completely logical. The input parameter is either a property name or an input parameter name, so we use the least specific name here.

File: modules/tensor_mechanics/src/materials/ScalarDamageBase.C [hunk 4]
```diff
@@ -30,16 +30,16 @@ ScalarDamageBase::validParams()
...
                                "name of the material property where the damage index is stored");
   return params;
 }
 
 ScalarDamageBase::ScalarDamageBase(const InputParameters & parameters)
   : DamageBase(parameters),
-    _damage_index_name(getParam<std::string>("damage_index_name")),
-    _damage_index(declareProperty<Real>(_base_name + _damage_index_name)),
+    _damage_index_name(getParam<MaterialPropertyName>("damage_index_name")),
+    _damage_index(declarePropertyByName<Real>(_base_name + _damage_index_name)),
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/ScalarDamageBase.C:42]: This change would need to be made in a million other classes as well.

Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/ScalarDamageBase.C:42]: ...? This is the only one that popped up in modules and apps?

Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/ScalarDamageBase.C:42]: It's typically fine everywhere else. the `ByName` is only required when there is a weird input params conflict.

Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/ScalarDamageBase.C:42]: `declarePropertyByName` bypasses the automatic name resolution from input parameters. Similar to `getMaterialProperty` and `getMaterialPropertyByName`...

Comment (dschwen) [reply modules/tensor_mechanics/src/materials/ScalarDamageBase.C:43]: Yeah, I'm aware of how that works. I also appreciate this API homogenization. I also think that if the developer knows they are passing in a name they should _always_ use the ByName method, and I'd actually prefer if there wasn't that voodoo magic that behind the scenes allows the argument to mean two entirely different things.

It's just confusing.

Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/ScalarDamageBase.C:43]: The "homogenization" is what I'm going for. I don't mind either way, but it's really confusing when `get` and `declare` do different things.

So you just griping or proposing a different way :)

Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/ScalarDamageBase.C:43]: Unless you're proposing to replace all existing `declareProperty<T>("some_name")` with `declarePropertyByName<T>("some_name")`, and `declareProperty<T>(getParam<MaterialPropertyName>("some_param"))` with `declareProperty<T>("some_param")`....

Comment (dschwen) [reply modules/tensor_mechanics/src/materials/ScalarDamageBase.C:43]: That would be my preference. But that's a bit beyond the scope of this PR. I'd love if we could deprecate the use of `declareProperty` and `getMaterialProperty` where the argument is _not_ a valid input parameter name.

Comment (tophmatthews) [reply modules/tensor_mechanics/src/materials/ScalarDamageBase.C:43]: Well, this PR has certainly come a loooooong way from it's original doomsday beginnings....

**Comment by tophmatthews** [comment]
So what's the verdict here @dschwen?

**Review by dschwen — COMMENTED** [review]
I'm fine with this, but it is missing the old and older versions

**Comment by dschwen** [comment]
Could we get another opinion from @idaholab/moose-team ?

**Comment by aeslaughter** [comment]
@dschwen Can you sum what is trying to be accomplished?

**Comment by dschwen** [comment]
getMaterialProperty accepts either an input parameter name or a material
property name.
This PR makes declareProperty behave the same way.

It is not touching the old and older versions, and I think in general this
dual functionality is confusing, but at least it is making the behavior
between declare and get more uniform.

**Comment by aeslaughter** [comment]
I like the consistency, but agree with @dschwen that the dual purpose is confusing. I am good with merging this, but should we quickly implement something like the following.

`getMaterialProperty`: uses parameter name if found (correct behavior) and issue a deprecated warning to use  getMatrerialPropertyByName if parameter it not found

Do the same for declareProperty

**Comment by joshuahansel** [comment]
I second @aeslaughter's comment.

**Comment by tophmatthews** [comment]
> getMaterialProperty: uses parameter name if found (correct behavior) and issue a deprecated warning to use getMatrerialPropertyByName if parameter it not found

Sounds easy enough. I can search through the framework and modules here to do an initial `declareProperty<T>("some_name")` --> `declareProperty<T>("some_name")` and `declareProperty<T>(getParam<MaterialPropertyName>("some_other_name"))` --> `declareProperty<T>("some_other_name")` here if wanted.

> It is not touching the old and older versions,

Do you want me to add `declarePropertyOldByName` even though `declarePropertyOld` is deprecated?

**Comment by dschwen** [comment]
> Do you want me to add `declarePropertyOldByName` even though `declarePropertyOld` is deprecated?

Ugh, no. I that was a dumb comment, sorry. Forget that I said anything :-D

**Comment by tophmatthews** [comment]
Maybe worth saving the many changes for a different pr...

**Review by dschwen — APPROVED** [review]

**Comment by lindsayad** [comment]
> `getMaterialProperty`: uses parameter name if found (correct behavior) and issue a deprecated warning to use getMatrerialPropertyByName if parameter it not found

Whose fault is this though? Is it the input file writer or the material writer? I think use of `getMaterialPropertyByName` in `Material` code is very inflexible and people shouldn't do it.

**Comment by lindsayad** [comment]
Well I guess if the `Material` writer wants to have the flexibility door open, they should have two things:

1. get(Required)Param<MaterialPropertyName>
2. geMaterialProperty

So perhaps it does make sense to lay the deprecation warning at the material writer's feet because if they don't make the param required, then they should supply a default value for the material property name.

**Pull request merged**